### PR TITLE
fixes #2406: scanner factory deadlock

### LINKED ISSUE #2406: scanner deadlock
We discovered a situation where we can deadlock the scanner factory.  The situation occurs in version 7.x and later:
Thread 1:
    ScannerFactory is synchronized in the close method (line 271)
    This calls BatchScannerSession.close
    Code was added to await terminations of the threads without a timeout (line 726)

Thread 2:
    The RangeStreamScanner (which is a scanner), calls close on the ScannerFactory as well in findTop.
    The problem is that findTop is in a thread that is now being awaited termination in thread 2

### PR DESCRIPTION
**Pull request by ivakegg**
* Removed synchronization in the scanner factory

### CODE REVIEW

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java [hunk 1]
```diff
@@ -33,36 +34,39 @@
...
+    protected final AccumuloClient cxn;
+    // using an AtomicBoolean to give us a separate monitor for synchronization
+    protected final AtomicBoolean open = new AtomicBoolean(true);
+
+    protected boolean accrueStats;
     protected Query settings;
-    protected ResourceQueue scanQueue = null;
-    ShardQueryConfiguration config = null;
+    protected ResourceQueue scanQueue;
+    ShardQueryConfiguration config;
```
Comment (billoley) [inline warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]:
```suggestion
    protected ShardQueryConfiguration config;
```

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java [hunk 2]
```diff
@@ -72,52 +76,75 @@ public ScannerFactory(AccumuloClient client) {
...
             Scanner bs = QueryScannerHelper.createScannerWithoutInfo(cxn, tableName, auths, query);
             log.debug("Created scanner " + System.identityHashCode(bs));
             if (log.isTraceEnabled()) {
                 log.trace("Adding instance " + bs.hashCode());
             }
 
-            return bs;
+            synchronized (open) {
+                if (open.get()) {
+                    return bs;
```
Comment (billoley) [inline warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java:93]: Shouldn't we add bs to instances here?
instances.add(bs);

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java [hunk 3]
```diff
@@ -213,22 +246,26 @@ public synchronized <T extends ScannerSession> T newLimitedScanner(Class<T> wrap
...
-        if (removed) {
+    public boolean close(ScannerBase bs) {
+        boolean removed = false;
+        try {
             log.debug("Closed scanner " + System.identityHashCode(bs));
+            removed = instances.remove(bs);
             if (log.isTraceEnabled()) {
                 log.trace("Closing instance " + bs.hashCode());
             }
             bs.close();
```
Comment (billoley) [inline warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]:
```suggestion
            if (removed) {
                if (log.isTraceEnabled()) {
                    log.trace("Closing instance " + bs.hashCode());
                }
                bs.close();
            }
```

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java [hunk 4]
```diff
@@ -247,21 +284,22 @@ public synchronized Collection<ScannerBase> currentScanners() {
...
+        synchronized (open) {
+            return open.getAndSet(false);
+        }
     }
 
-    public synchronized void close(ScannerSession bs) {
+    public void close(ScannerSession bs) {
         try {
             log.debug("Closed session " + System.identityHashCode(bs));
             sessionInstances.remove(bs);
```
Comment (billoley) [inline warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]: check if removed?

**Review by billoley — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java (hunk 1 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]: agreed... done

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java (hunk 2 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java:93]: agreed... done.

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java (hunk 3 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]: agreed.... done.

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java (hunk 4 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]: agreed.... done.

**Review by billoley — APPROVED** [review]

**Review by keith-ratcliffe — APPROVED** [review]

**Pull request merged**